platform-sdk/swirlds-common/src/main/java/com/swirlds/common/wiring/internal/MeteredSequentialWire.java

```diff
@@ -0,0 +1,207 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.common.wiring.internal;
+
+import com.swirlds.common.metrics.extensions.FractionalTimer;
+import com.swirlds.common.metrics.extensions.NoOpFractionalTimer;
+import com.swirlds.common.wiring.Wire;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+/**
+ * A {@link Wire} that guarantees that tasks are executed sequentially in the order they are received.
+ *
+ * @param <T> the type of object that is passed through the wire
+ */
+public class MeteredSequentialWire<T> implements Wire<T> {
+    private Consumer<T> consumer;
+    private final AtomicReference<SequentialTask> lastTask;
+    private final AbstractObjectCounter counter;
+    private final FractionalTimer busyTimer;
+    private final String name;
+
+    /**
+     * A task in a sequential wire.
+     *
+     */
+    private class SequentialTask extends AbstractTask {
+        private T data;
+        private SequentialTask nextTask;
+
+        /**
+         * Constructor.
+         *
+         * @param dependencyCount the number of dependencies that must be satisfied before this task is eligible for
+         *                        execution. The first task in a sequence has a dependency count of 1 (data must be
+         *                        provided), and subsequent tasks have a dependency count of 2 (the previous task must
+         *                        be executed and data must be provided).
+         */
+        SequentialTask(final int dependencyCount) {
+            super(dependencyCount);
+        }
+
+        /**
+         * Provide a reference to the next task and the data that will be processed during the handling of this task.
+         *
+         * @param nextTask the task that will execute after this task
+         * @param data     the data to be passed to the consumer for this task
+         */
+        void send(@NonNull final SequentialTask nextTask, @NonNull final T data) {
+            this.nextTask = nextTask;
+            this.data = data;
+
+            // This method provides us with the data we intend to send to the consumer
+            // when this task is executed, thus resolving one of the two dependencies
+            // required for the task to be executed. This call will decrement the
+            // dependency count. If this causes the dependency count to reach 0
+            // (i.e. if the previous task has already been executed), then this call
+            // will cause this task to be immediately eligible for execution.
+            send();
+        }
+
+        /**
+         * Execute this task.
+         */
+        @Override
+        public boolean exec() {
+            counter.offRamp();
+            busyTimer.activate();
```

Comment:
`SequentialTask`s are executed serially by design. Paying high price for updating atomics is not justified here (applies to `busyTimer` but not `counter`).